src/haddock/modules/analysis/contactmap/contmap.py

```diff
@@ -0,0 +1,708 @@
+        matrix: NDFloat,
+        indices: list[int],
+        indices2: Optional[list[int]] = None,
+        ) -> NDFloat:
+    """Extract submatrix based on desired indices.
+
+    Paramaters
+    ----------
+    matrix : NDFloat
+        A N*N matrix.
+    indices : list[int]
+        List of `row` indices to extract from this matrix
+    indices2 : list[int]
+        List of `columns` indices to extract from this matrix.
+         if unspecified, indices2 == indices and symetric matrix
+         is extracted.
+
+    Return
+    ------
+    submat : NDFloat
+        The extracted submatrix.
+    """
+    # Set second set of indices (columns) to first if not defined
+    if not indices2:
+        indices2 = indices
+    # extract submatrix
+    submat = matrix[np.ix_(indices, indices2)]
+    return submat
+
+
+def gen_contact_dt(
+        matrix: NDFloat,
+        resdt: dict,
+        res1_key: str,
+        res2_key: str,
+        ) -> dict:
+    """Generate contacts data.
+
+    Parameters
+    ----------
+    matrix : NDFloat
+        The distance matrix.
+    resdt : dict
+        Residues data with atom indices as returned by `get_ordered_coords()`.
+    res1_key : str
+        First residue of interest.
+    res2_key : str
+        Second residue of interest
+
+    Return
+    ------
+    cont_dt : dict
+        Dictionary holding contact data
+    """
+    # point residues data
+    res1_dt = resdt[res1_key]
+    res2_dt = resdt[res2_key]
+    # point ca-ca dist
+    try:
+        ca_ca_dist = matrix[res1_dt['CA'], res2_dt['CA']]
+    except KeyError:
+        ca_ca_dist = 9999
+    # obtain submatrix
+    res1_res2_atm_submat = extract_submatrix(
+        matrix,
+        res1_dt['atoms_indices'],
+        res2_dt['atoms_indices'],
+        )
+    # obtain clostest contact
+    clostest_contact = min_dist(res1_res2_atm_submat)
+    # contact type
+    cont_type = get_cont_type(res1_dt['resname'], res2_dt['resname'])
+    # set return variable
+    cont_dt = {
+        'res1': res1_key,
+        'res2': res2_key,
+        'ca-ca-dist': round(ca_ca_dist, 1),
+        'shortest-dist': round(clostest_contact, 1),
+        'contact-type': cont_type,
+        # add other types of data ??
+        }
+    return cont_dt
+
+
+def get_cont_type(resn1: str, resn2: str) -> str:
+    """Generate polarity key between two residues.
+
+    Parameters
+    ----------
+    resn1 : str
+       3 letters code of fist residue.
+    resn2 : str
+       3 letters code of second residue.
+    
+    Return
+    ------
+    pol_key : str
+        Combined residues polarities
+    """
+    pol_key = f'{to_res_polarity(resn1)}-{to_res_polarity(resn2)}'
+    return pol_key
+
+
+def to_res_polarity(resn: str) -> str:
+    """Convert residue name into polarity.
+
+    Parameters
+    ----------
+    resn : str
+       3 letters code of a residue.
+    """
+    try:
+        return RESIDUE_POLARITY[resn]
+    except KeyError:
+        return 'unknow'
```

Comment:
Making a function for a try/except block that uses a global variable is messy - why not have this in the code block?